KAFKA-13152: Replace "buffered.records.per.partition" with "input.buffer.max.bytes"

vamossagar12 opened the pull request:
This PR is an implementation of: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=186878390. The following changes have been made:

- Adding a new config input.buffer.max.bytes applicable at a topology level. 
- Adding new config statestore.cache.max.bytes. 
- Adding new metric called total-bytes .
- The per partition config buffered.records.per.partition is deprecated.
- The config  cache.max.bytes.buffering is deprecated.

Commit:
KAFKA-13152: Initial commit

Commit:
KAFKA-13152: Initial commit

Commit:
Adding more tests

Commit:
Adding more tests

Commit:
Adding test cases and fixing check style issues

Commit:
Adding test cases and fixing check style issues

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1030,8 +1038,11 @@ private static Metrics getMetrics(final StreamsConfig config, final Time time, f
                     threads.remove(streamThread);
                     final long cacheSizePerThread = getCacheSizePerThread(getNumLiveStreamThreads());
+                    final long maxBufferSizePerThread = getBufferSizePerThread(getNumLiveStreamThreads());
                     log.info("Resizing thread cache due to terminating added thread, new cache size per thread is {}", cacheSizePerThread);
```
The phrasing here is a little awkward, can we try to clean it up a bit here and in the new line below? (Also worth considering to combine into a single log line here, although in this case imo it's fine either way)
```suggestion
                    log.info("Resizing thread cache again since new thread can not be started, final cache size per thread is {}", cacheSizePerThread);
```

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1221,13 +1236,27 @@ private long getCacheSizePerThread(final int numStreamThreads) {
         return totalCacheSize / (numStreamThreads + (topologyMetadata.hasGlobalTopology() ? 1 : 0));
     }
 
+    private long getBufferSizePerThread(final int numStreamThreads) {
```
Seems like this is exactly the same logic as `getCacheSizePerThread`, can we instead just change the name of the existing method to match both cases rather than writing duplicate code? Maybe something like `getMemorySharePerThread` or `getMemorySizePerThread`?

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1013,12 +1017,16 @@ private static Metrics getMetrics(final StreamsConfig config, final Time time, f
                 log.info("Adding StreamThread-{}, there will now be {} live threads and the new cache size per thread is {}",
                          threadIdx, numLiveThreads + 1, cacheSizePerThread);
                 resizeThreadCache(cacheSizePerThread);
+                log.info("Adding StreamThread-{}, there will now be {} live threads and the new buffer size per thread is {}",
```
Seems kind of weird to log this twice, it makes it seem like we're adding this StreamThread a second time. How about instead just adding the new buffer size to the original log message (eg something like)
```suggestion
                log.info("Adding StreamThread-{}, there are now {} threads with a buffer size {} and cache size {} per thread.",
```

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1221,13 +1236,27 @@ private long getCacheSizePerThread(final int numStreamThreads) {
         }
     }
 
+    private void resizeMaxBufferSize(final long maxBufferSize) {
```
Same here, if it's going to be exactly the same then we only need one method.

Although, I don't think the GlobalThread actually even has an input buffer the way StreamThreads do (doesn't seem to make sense for it to need one because it can just process all of its polled records right away, whereas the StreamThread may need to buffer them for various reasons)

You could still probably combine into a single method and just include a flag for whether or not to call `resize` on the global thread (with it being true for the cache case, and false for the input buffer resizing)

guozhangwang replied:
+1 here as well. I think we would always resize both buffer and state cache at the same time moving forward.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -747,11 +764,6 @@
                     "",
                     Importance.LOW,
                     APPLICATION_SERVER_DOC)
-            .define(BUFFERED_RECORDS_PER_PARTITION_CONFIG,
```
I think you still need (or at least want) to keep these deprecated configs defined

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -385,7 +385,7 @@ public synchronized final InternalTopologyBuilder rewriteTopology(final StreamsC
 
         // maybe strip out caching layers
-        if (config.getLong(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG) == 0L) {
+        if (config.getLong(StreamsConfig.STATESTORE_CACHE_MAX_BYTES_CONFIG) == 0L) {
```
The original config is deprecated but by definition may still be in use, so we need to check both configs. Also need to consider what to do if both are defined (imo just use whatever the value for the new config and ignore the old one, but you could log a warning if both are set to different values)

vamossagar12 replied:
Yeah. I wasn't quite sure whether I should rename everywhere else or not. I chose to do it. Will add it back- will also make the number of files changed a lot less.

ableegoldman replied:
Ah I wasn't trying to suggest we shouldn't start using the new config name everywhere, we absolutely should (even if it does explode the PR quite a bit -- I'll live). My point here was that we need to check the value of both of these configs, since we don't know which (if any) the user actually set.

Which, by the way,  you can check via `config.originals()` -- this returns the set of configs the user passed in, otherwise it's tricky to figure out since eg `config.getLong` would return the config default value if not set

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java:
```diff
@@ -47,6 +47,7 @@
     private final ArrayDeque<ConsumerRecord<byte[], byte[]>> fifoQueue;
 
     private StampedRecord headRecord = null;
+    private long headRecordSizeInBytes = 0;
```
Is this used anywhere?

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -721,6 +732,10 @@ public void resizeCache(final long size) {
         cacheResizeSize.set(size);
     }
 
+    public void resize(final long size) {
```
nit: call this one `resizeCache` to differentiate between this and the buffer size? Also, is there a method somewhere to update the input buffer size? Seems like we're always resizing the cache when a thread is added/removed

vamossagar12 replied:
Haha yeah that's a good catch. Missed that part. This is the method which is supposed to do the same

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -601,6 +608,10 @@ boolean runLoop() {
                 if (size != -1L) {
                     cacheResizer.accept(size);
                 }
+                final long bufferBytesSize = maxBufferResizeSize.getAndSet(-1L);
```
Do we need this? Seems like we can get rid of the extra `maxBufferSizeBytes` and instead just directly read out the value of `maxBufferResizeSize` (in which case we should probably rename `maxBufferResizeSize`  to `maxBufferSizeBytes` -- my point is, I think we can get away with just storing the current size of the buffer as a `long` and the max size as an `AtomicLong` that can be updated from outside the thread)

vamossagar12 replied:
> Thanks for the PR! Lmk if you have any questions about any of my comments/questions -- I think the main thing is just that we, unfortunately, can't get rid of any of the code handling the old configs just yet, since not everyone will upgrade their code off of deprecated APIs right away (though we wish they would!)

Thanks for the review. I have added some comments/questions. Other suggestions, make total sense and would incorporate them.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -793,6 +809,10 @@ void runOnce() {
 
                     totalProcessed += processed;
                     totalRecordsProcessedSinceLastSummary += processed;
+                    if (bufferSize > maxBufferSizeBytes && bufferSize - processedData.totalBytesConsumed <= maxBufferSizeBytes) {
```
This logic is a little difficult to read 😅  -- would it help to just move the `bufferSize -= processedData.totalBytesConsumed;` to before this check?

Although on that note, it might be cleaner to just move this check regarding whether to resume consuming to right before we call `poll`, that way there's a nice symmetry between the `pause` and `resume` checks, and all the logic is consolidated to one place

vamossagar12 replied:
Haha :D 
The reason I have this check => `bufferSize > maxBufferSizeBytes` is that the resume should happen only if after the current round of consumption, the buffersize which had breached the threshold, now went below. Without that check, it will always enter the if block- even when it's already lesser(and we subtract something more and reduce it further). Did that make sense? :D 
The idea of placing it here is that, right after removing some records from the buffer, did the buffer size come down. It similar to how StreamThread resume used to  work(the one I removed). This logic can very well go before poll but I thought adding it here was more non invasive as there's already some metrics related stuff and other things happening in this block.

guozhangwang replied:
I also feel this logic is a bit awkward, starting from the fact that we need to report how many bytes we've consumed from the process :) I think we can simply do the following:

At the end of polling phase, and at the end the process loop (a.k.a. here), we loop over all the active tasks, and get their "input buffer size", which would delegate to each task's corresponding `PartitionGroup` and then `RecordQueue`. And then based on that we can decide whether to resume / pause accordingly. Then

1) we do not need to maintain a local `bufferSize` at the task here, i.e. we always re-compute from the task's record queue, which is the source of truth.
2) we do not need to maintain and propagate up the `consumed bytes` within each iteration here.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1057,6 +1057,17 @@ void addRecordsToTasks(final ConsumerRecords<byte[], byte[]> records) {
     }
 
+    /**
+     *  Fetch all non-empty partitions for pausing
```
Why do we only pause the non-empty partitions? If the buffer is full, we have to pause all of them, no?

vamossagar12 replied:
Yeah.. this was something that was discussed/decided in the JIRA conversation. You can find the explanation here: https://issues.apache.org/jira/browse/KAFKA-13152?focusedCommentId=17400647&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17400647

ableegoldman replied:
Ah now I remember, thanks 😄

ableegoldman replied:
Can you leave a comment above this method explaining this? Don't want to forget why we did this a year from now and then accidentally break things

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -736,11 +734,8 @@ public boolean process(final long wallClockTime) {
             consumedOffsets.put(partition, record.offset());
             commitNeeded = true;
 
-            // after processing this record, if its partition queue's buffered size has been
-            // decreased to the threshold, we can then resume the consumption on this partition
-            if (recordInfo.queue().size() == maxBufferedSize) {
-                mainConsumer.resume(singleton(partition));
-            }
```
Unfortunately we can't remove this logic yet, since some users may still be setting only the `buffered.records.per.partition` config, and they shouldn't see a sudden explosion of memory just because they didn't switch over to the new config right away

vamossagar12 replied:
I see... but if both the configs are in play, won't that also lead to pause/resume of partitions based upon whichever threshold is breached? Maybe we can add some check that we do this only if maxBufferedSize is set to some value? We might also want to consider if it has a default value and use the condition accordingly,

ableegoldman replied:
>  Maybe we can add some check that we do this only if maxBufferedSize is set to some value?

Yeah, sorry, I should have been more clear here -- we only need to continue doing this if the user is still setting the `buffered.records.per.partition` config. I mentioned this in another comment, but just in case you weren't aware, you can call `originals()` on a StreamsConfig object to get a map of the actual configs passed in by the user -- that way you know what they actually set vs what's just a default.

Then you can just set `maxBufferedSize` to null or define a static constant `NOT_SET = -1` and then only continue doing this partition-level pause/resume if the user is still using `buffered.records.per.partition`. Does that make sense?

ableegoldman reviewed (commented):
Thanks for the PR! Lmk if you have any questions about any of my comments/questions -- I think the main thing is just that we, unfortunately, can't get rid of any of the code handling the old configs just yet, since not everyone will upgrade their code off of deprecated APIs right away (though we wish they would!)

vamossagar12 committed:
Update streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

Co-authored-by: A. Sophie Blee-Goldman <ableegoldman@gmail.com>

vamossagar12 committed:
Update streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

Co-authored-by: A. Sophie Blee-Goldman <ableegoldman@gmail.com>

vamossagar12 committed:
Update streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

Co-authored-by: A. Sophie Blee-Goldman <ableegoldman@gmail.com>

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1013,12 +1017,16 @@ private static Metrics getMetrics(final StreamsConfig config, final Time time, f
                 log.info("Adding StreamThread-{}, there will now be {} live threads and the new cache size per thread is {}",
                          threadIdx, numLiveThreads + 1, cacheSizePerThread);
                 resizeThreadCache(cacheSizePerThread);
+                log.info("Adding StreamThread-{}, there are now {} threads with a buffer size {} and cache size {} per thread.",
```
nit: Could we merge these two info lines into a single one? It seems a bit redundant to log twice here. Also this new log line seems wrong since it has four parameters but only three values provided.

E.g.

```
Adding StreamThread-{}, the current total number of thread is {}, each thread now has a buffer size {} and cache size {}
```

And

```
Terminating StreamThread-{}, the current total number of thread is {}, each thread now has a buffer size {} and cache size {}
```

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1030,8 +1038,11 @@ private static Metrics getMetrics(final StreamsConfig config, final Time time, f
                     final long cacheSizePerThread = getCacheSizePerThread(getNumLiveStreamThreads());
-                    log.info("Resizing thread cache due to terminating added thread, new cache size per thread is {}", cacheSizePerThread);
+                    final long maxBufferSizePerThread = getBufferSizePerThread(getNumLiveStreamThreads());
+                    log.info("Resizing thread cache again since new thread can not be started, final cache size per thread is {}", cacheSizePerThread);
```
nit: ditto here. See above for the consolidated log line. Here we can emphasize it is "Terminating newly added threads".

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1112,6 +1124,9 @@ private static Metrics getMetrics(final StreamsConfig config, final Time time, f
                         log.info("Resizing thread cache due to thread removal, new cache size per thread is {}", cacheSizePerThread);
                         resizeThreadCache(cacheSizePerThread);
+                        final long maxBufferSizePerThread = getBufferSizePerThread(getNumLiveStreamThreads());
+                        log.info("Resizing max buffer size due to thread removal, new buffer size per thread is {}", maxBufferSizePerThread);
```
Ditto here as wel.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1269,20 +1280,36 @@ void handleTopologyUpdates() {
         for (final Task task : activeTaskIterable()) {
             int processed = 0;
+            long bytesConsumed = 0L;
+            task.setBytesConsumed(0L);
```
Not sure I understand this logic here: it seems we only call `setBytesConsumed` once with 0 here and there's no other callers?

guozhangwang replied:
Also, even with the correct logic, I'm wondering if we can just define it as a local variable within the `process` here instead of augmenting the `Task` interface?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java:
```diff
@@ -266,6 +270,10 @@ StampedRecord nextRecord(final RecordInfo info, final long wallClockTime) {
             if (record != null) {
                 --totalBuffered;
 
+                totalBytesBuffered -= (record.key() != null ? record.serializedKeySize() : 0) +
```
How about adding a `sizeInBytes` function to `StampedRecord` class, to avoid duplicating the calculation each time?

Also we do not need to check for `key/value == null` or not, since in that case the `ConsumerRecord#serializedKeySize/ValueSize` would be 0.

Also note that the total bytes taken here is more than just the key/value since we have other fields like theaders, timestamps etc, but since it's hard to capture exactly how many bytes are taken in JVM to store a ConsumerRecord<byte[], byte[]> I think getting just getting the sum of key/value/timestamp/offset/topic/partition/headers is fine.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java:
```diff
@@ -354,10 +370,22 @@ int numBuffered(final TopicPartition partition) {
         return totalBuffered;
     }
 
+    long totalBytesBuffered() {
```
This function seems only used for testing? If yes, please move it to the bottom of the class and add a comment that "below are for testing only".

guozhangwang replied:
This function seems not addressed? Or was I missing anything?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StampedRecord.java:
```diff
@@ -49,6 +49,14 @@ public Headers headers() {
         return value.headers();
     }
 
+    public int serializedKeySize() {
```
Please see my other comment above: how about just have a `sizeInBytes` function which takes key/value/timestamp/offset/topic/partition/headers into account?

guozhangwang replied:
These functions seem not used any more?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/TaskMetrics.java:
```diff
@@ -84,6 +84,9 @@ private TaskMetrics() {}
         "from consumer and not yet processed for this active task";
 
+    private static final String TOTAL_BYTES = "total-bytes";
+    private static final String TOTAL_BYTES_DESCRIPTION = "The total number of bytes accumulated by this task";
```
We should be more specific about the description here: the total number of bytes accumulated in this task's input buffer.

guozhangwang replied:
This comment seems not addressed.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/TaskMetrics.java:
```diff
@@ -128,6 +131,22 @@ public static Sensor activeBufferedRecordsSensor(final String threadId,
         return sensor;
     }
 
+    public static Sensor totalBytesSensor(final String threadId,
```
While reviewing the PR, I feel the name `total-bytes` under

```
type = stream-task-metrics
thread-id = [thread ID]
task-id = [task ID]
```

is a bit too vague, what about renaming it to `total-input-buffer-bytes`. WDYT @vamossagar12 @ableegoldman ? If we agree here we'd need to update the KIP as well.

vamossagar12 replied:
That makes sense to me. WDYT @ableegoldman ?

vamossagar12 replied:
Btw I am more or less done with all other changes. So if we can decide on this, I would update it. There was another thing which I had missed with update partitions in partition group where if a partition is removed, we needed to remove the corresponding buffer size of it.

guozhangwang replied:
cc @ableegoldman let me know WDYT about the new metric name?

vamossagar12 replied:
hey @ableegoldman , the list of conflicting files is growing on this PR :D Whenever you get the chance could you plz share your thoughts on this and a couple of other PRs ([this](https://github.com/apache/kafka/pull/11211) and [this](https://github.com/apache/kafka/pull/11433)

vamossagar12 replied:
hi.. bumping this thread again

guozhangwang replied:
@ableegoldman ping :)

@vamossagar12 I'm just back from a long vacation and is reloading all my review memory here :) is this ready for another review now besides this comment for naming? If yes could you rebase and ping me?

vamossagar12 replied:
@guozhangwang Welcome back :) From my side, yeah I think it's ready for another review besides the comment. I have rebased the changes as well.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -601,6 +608,10 @@ boolean runLoop() {
                     cacheResizer.accept(size);
                 }
+                final long bufferBytesSize = maxBufferResizeSize.getAndSet(-1L);
+                if (size != -1) {
```
This does not look right to me: why we use the size value read from `cacheResizeSize` to assign to `maxBufferSizeBytes`? They should be totally orthogonal.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1269,20 +1280,36 @@ void handleTopologyUpdates() {
-    int process(final int maxNumRecords, final Time time) {
+    RecordsProcessedMetadata process(final int maxNumRecords, final Time time) {
         int totalProcessed = 0;
+        long totalBytesConsumed = 0L;
```
I think we can simply the logic and do not need to keep track of "consumed bytes" within a task here, see my other comment.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1288,6 +1326,10 @@ int currentNumIterations() {
         return numIterations;
     }
 
+    long bufferSize() {
```
This function seems not used.

BTW if we do not maintain the local `bufferSize` then we would not need it anyways :)

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1269,20 +1280,36 @@ void handleTopologyUpdates() {
         }
     }
 
+    static class RecordsProcessedMetadata {
```
See my other comment: I think we can avoid propagating both processed-records and processed-bytes from the `process` call.

guozhangwang reviewed (commented):
Thanks for the PR! I made a first pass on it.

Commit:
Fixing test cases and correcting checkstyle issues

Commit:
Adding missed files due to rebase

Commit:
Rebasing with trunk

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1074,9 +1083,22 @@ private static Metrics getMetrics(final StreamsConfig config, final Time time, f
      * The removed stream thread is gracefully shut down. This method does not specify which stream
      * thread is shut down.
      * <p>
+<<<<<<< HEAD
```
Rebase leftovers here.

vamossagar12 replied:
Removed

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1116,16 +1138,18 @@ private static Metrics getMetrics(final StreamsConfig config, final Time time, f
+                        resizeThreadCacheOrBufferMemory(cacheSizePerThread, true);
+                        final long maxBufferSizePerThread = getMemorySizePerThread(getNumLiveStreamThreads(), false);
+                        resizeThreadCacheOrBufferMemory(maxBufferSizePerThread, false);
+                        log.info("Resizing thread cache/max buffer size due to thread removal, new cache size/max buffer size per thread is {}/{}", cacheSizePerThread, maxBufferSizePerThread);
```
Please see my other comment earlier: I'd suggest we also include the terminating thread name in the log message here for better trouble-shooting, this would be valuable when multiple threads are being created / terminated at the same time.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1224,17 +1248,20 @@ private int getNextThreadIndex() {
     }
 
-    private long getCacheSizePerThread(final int numStreamThreads) {
+    private long getMemorySizePerThread(final int numStreamThreads, final boolean resizeCache) {
```
I think what's originally suggested is a bit more than having a single consolidated function, but also to consolidate the caller: since we are always calling "getMemorySizePerThread" and `resizeThreadCacheOrBufferMemory` for both input buffer and cache size at the same place, we can have a single call rather than two calls. For example, we can just have:

1) a single `resizeThreadCacheAndBufferMemory(final long numThreads)` in which we get the cache-size and input-buffer-size per thread based on the total values of the two and the passed in number of threads, and then let each thread set to the corresponding values directly.

2) the `getMemorySizePerThread` function is then only needed for logging purposes, since otherwise it is now encapsulated inside the `resizeThreadCacheAndBufferMemory` function. We can have just a single `getThreadCacheAndBufferMemoryString()` which returns in the format of "value1/value2" and used in the logging message alone; here we can just get any thread and get its values.

vamossagar12 replied:
@guozhangwang , For the most part this should work. However, in `addStreamThread` , when we need to add a new StreamThread, then at that point, the `createAndAddStreamThread` method expects the newly computed cache size and the new logic for max buffer size. With the consolidated logic and getMemorySizePerThread being converted to return a string to be used for logging purpose only, getting this value to be passed to the constructor might become difficult. Or am I missing something?

guozhangwang replied:
Could we create the new thread with `0` initial values first, and then call `resizeThreadCacheOrBufferMemory` to set the new values for all threads, including the new thread?

vamossagar12 replied:
Got it. I added the relevant logic.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -738,7 +739,7 @@ public boolean process(final long wallClockTime) {
             // after processing this record, if its partition queue's buffered size has been
             // decreased to the threshold, we can then resume the consumption on this partition
-            if (recordInfo.queue().size() == maxBufferedSize) {
+            if (maxBufferedSize != -1 && recordInfo.queue().size() == maxBufferedSize) {
```
nit: could we add a TODO here that this logic should be removed once we remove the deprecated old config as well?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -967,6 +980,17 @@ private long pollPhase() {
+                // Only non-empty partitions are paused here. Reason is that, if a task has multiple partitions with
+                // some of them empty, then in that case pausing even empty partitions would sacrifice ordered processing
+                // and even lead to temporal deadlock. More explanation can be found here:
+                // https://issues.apache.org/jira/browse/KAFKA-13152?focusedCommentId=17400647&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17400647
```
nit: we do not need the `?focusedCommentId=17400647&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17400647` suffix in the javadoc :)

vamossagar12 replied:
Oh that's because the ticket has a lot of comments and I intended to point to the comment which talks about the design decision made here. If it doesn't make sense, will remove it :)

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -728,8 +733,11 @@ private void subscribeConsumer() {
 
-    public void resizeCache(final long size) {
-        cacheResizeSize.set(size);
+    public void resizeCacheOrBufferMemory(final long size, final boolean cacheResize) {
```
See the other comment: if we always size the two at the same time, we can just make one function call instead of two calls with a second boolean flag.

guozhangwang reviewed (commented):
Thanks @vamossagar12 , I've made a second pass and left some more comments.

Regarding the metric name, I will ping @ableegoldman and @cadonna again and get back to you.

vamossagar12 commented:
https://github.com/apache/kafka/pull/11424#discussion_r773499306

This is being used now. StreamThread delegates to this function to get the totalBytesBuffered.

Commit:
Further Review comments

vamossagar12 commented:
> Thanks @vamossagar12 , I've made a second pass and left some more comments.
> 
> Regarding the metric name, I will ping @ableegoldman and @cadonna again and get back to you.

Thanks @guozhangwang . I had one question on the changes suggested by you.

guozhangwang commented:
> Thanks @guozhangwang . I had one question on the changes suggested by you.

Just replied on your question. LMK if that works or not.

Commit:
Adding logic to first create new StreamThread with 0 cache size and max buffer bytes size and then resizing

vamossagar12 commented:
> > Thanks @guozhangwang . I had one question on the changes suggested by you.
> 
> Just replied on your question. LMK if that works or not.

Yeah it does.. Made the changes. Plz let me know when we have the decision on the metric name. Also, plz review if any other changes are needed.

guozhangwang commented:
Thanks @vamossagar12 , I do not have further comments now, re-triggering the jenkins tests.

guozhangwang commented:
@vamossagar12 there are some compilation errors in jenkins, could you rebase from latest trunk and check what's the issue? Please ping me once that's resolved.

vamossagar12 commented:
@guozhangwang , it's due to the usage of deprecated configs and gradle having a check against it.

```
[2022-01-04T23:35:30.575Z] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11424/streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/TopologyConfig.java:61: warning: [deprecation] CACHE_MAX_BYTES_BUFFERING_CONFIG in org.apache.kafka.streams.StreamsConfig has been deprecated

[2022-01-04T23:35:30.575Z]             .define(CACHE_MAX_BYTES_BUFFERING_CONFIG,

[2022-01-04T23:35:30.575Z]                     ^

[2022-01-04T23:35:30.575Z] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11424/streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/TopologyConfig.java:56: warning: [deprecation] BUFFERED_RECORDS_PER_PARTITION_CONFIG in org.apache.kafka.streams.StreamsConfig has been deprecated

[2022-01-04T23:35:30.575Z]              .define(BUFFERED_RECORDS_PER_PARTITION_CONFIG,

[2022-01-04T23:35:30.575Z]                      ^

[2022-01-04T23:35:30.575Z] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11424/streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/TopologyConfig.java:115: warning: [deprecation] BUFFERED_RECORDS_PER_PARTITION_CONFIG in org.apache.kafka.streams.StreamsConfig has been deprecated

[2022-01-04T23:35:30.575Z]         if (isTopologyOverride(BUFFERED_RECORDS_PER_PARTITION_CONFIG, topologyOverrides)) {

[2022-01-04T23:35:30.575Z]                                ^

[2022-01-04T23:35:30.575Z] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11424/streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/TopologyConfig.java:116: warning: [deprecation] BUFFERED_RECORDS_PER_PARTITION_CONFIG in org.apache.kafka.streams.StreamsConfig has been deprecated

[2022-01-04T23:35:30.575Z]             maxBufferedSize = getInt(BUFFERED_RECORDS_PER_PARTITION_CONFIG);

[2022-01-04T23:35:30.575Z]                                      ^

[2022-01-04T23:35:30.575Z] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11424/streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/TopologyConfig.java:117: warning: [deprecation] BUFFERED_RECORDS_PER_PARTITION_CONFIG in org.apache.kafka.streams.StreamsConfig has been deprecated

[2022-01-04T23:35:30.575Z]             log.info("Topology {} is overriding {} to {}", topologyName, BUFFERED_RECORDS_PER_PARTITION_CONFIG, maxBufferedSize);

[2022-01-04T23:35:30.575Z]                                                                          ^

[2022-01-04T23:35:30.575Z] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11424/streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/TopologyConfig.java:119: warning: [deprecation] BUFFERED_RECORDS_PER_PARTITION_CONFIG in org.apache.kafka.streams.StreamsConfig has been deprecated

[2022-01-04T23:35:30.575Z]             maxBufferedSize = globalAppConfigs.originals().containsKey(StreamsConfig.BUFFERED_RECORDS_PER_PARTITION_CONFIG)

[2022-01-04T23:35:30.575Z]                                                                                     ^

[2022-01-04T23:35:30.575Z] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11424/streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/TopologyConfig.java:120: warning: [deprecation] BUFFERED_RECORDS_PER_PARTITION_CONFIG in org.apache.kafka.streams.StreamsConfig has been deprecated

[2022-01-04T23:35:30.575Z]                     ? globalAppConfigs.getInt(StreamsConfig.BUFFERED_RECORDS_PER_PARTITION_CONFIG) : -1;

[2022-01-04T23:35:30.575Z]                                                            ^

[2022-01-04T23:35:30.575Z] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11424/streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/TopologyConfig.java:123: warning: [deprecation] CACHE_MAX_BYTES_BUFFERING_CONFIG in org.apache.kafka.streams.StreamsConfig has been deprecated

[2022-01-04T23:35:30.575Z]         if (isTopologyOverride(CACHE_MAX_BYTES_BUFFERING_CONFIG, topologyOverrides)) {

[2022-01-04T23:35:30.575Z]                                ^

[2022-01-04T23:35:30.575Z] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11424/streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/TopologyConfig.java:124: warning: [deprecation] CACHE_MAX_BYTES_BUFFERING_CONFIG in org.apache.kafka.streams.StreamsConfig has been deprecated

[2022-01-04T23:35:30.575Z]             cacheSize = getLong(CACHE_MAX_BYTES_BUFFERING_CONFIG);

[2022-01-04T23:35:30.575Z]                                 ^

[2022-01-04T23:35:30.575Z] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11424/streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/TopologyConfig.java:125: warning: [deprecation] CACHE_MAX_BYTES_BUFFERING_CONFIG in org.apache.kafka.streams.StreamsConfig has been deprecated

[2022-01-04T23:35:30.575Z]             log.info("Topology {} is overriding {} to {}", topologyName, CACHE_MAX_BYTES_BUFFERING_CONFIG, cacheSize);

[2022-01-04T23:35:30.575Z]                                                                          ^

[2022-01-04T23:35:30.575Z] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11424/streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/TopologyConfig.java:127: warning: [deprecation] CACHE_MAX_BYTES_BUFFERING_CONFIG in org.apache.kafka.streams.StreamsConfig has been deprecated

[2022-01-04T23:35:30.575Z]             cacheSize = globalAppConfigs.getLong(CACHE_MAX_BYTES_BUFFERING_CONFIG);

[2022-01-04T23:35:30.575Z]                                                  ^

[2022-01-04T23:35:30.575Z] error: warnings found and -Werror specified
```

In my local I commented out that check for compile/test/checkstyle etc. I think we don't want to get rid of these configs immediately, so these usages are needed for now. What would you suggest here?

guozhangwang commented:
@vamossagar12 Note that `TopologyConfig` is an internal class for now, and hence should not use deprecated values, I think we need to update this class to directly use the new config and remove the old ones --- since it is internal, it would not be breaking any compatibilities.

vamossagar12 commented:
hey @guozhangwang , For `CACHE_MAX_BYTES_BUFFERING_CONFIG` , I can replace but as per  one of the comments here => https://github.com/apache/kafka/pull/11424#discussion_r740701956 we wanted to check if the deprecated config (BUFFERED_RECORDS_PER_PARTITION_CONFIG) has been set or otherwise set it to -1. For setting it to -1, I am also checking if it's a user supplied value or a default one and that's where the need to use this config in TopologyConfig class.

Commit:
Trying to remove deprecated config

vamossagar12 commented:
> hey @guozhangwang , For `CACHE_MAX_BYTES_BUFFERING_CONFIG` , I can replace but as per one of the comments here => [#11424 (comment)](https://github.com/apache/kafka/pull/11424#discussion_r740701956) we wanted to check if the deprecated config (BUFFERED_RECORDS_PER_PARTITION_CONFIG) has been set or otherwise set it to -1. For setting it to -1, I am also checking if it's a user supplied value or a default one and that's where the need to use this config in TopologyConfig class.

@guozhangwang , i tried to address this again, but what I see is that this config BUFFERED_RECORDS_PER_PARTITION_CONFIG is being used to set maxBufferedSize only in TopologyConfig. As I said, for backward compatibility reasons, we are checking if it has been set by the user or not.. I know it's dragging a bit here but any suggestions would be welcome here :D

vamossagar12 commented:
hey @guozhangwang , did you get a chance to look at my above comment?

guozhangwang commented:
> hey @guozhangwang , did you get a chance to look at my above comment?

Ah yes, for StreamsConfigs in order for backward compatibility we still need to reference some deprecated configs. So:

1) For all references of `CACHE_MAX_BYTES_BUFFERING_CONFIG`, we should be replacing it with the new config. We just add the `@suppressWarning` in the place we reference it only for [compatibility](https://github.com/apache/kafka/pull/11424/files#diff-07a8a87d3ee84c3e96918aa131033780cf4bc4d4a12ae376e91c059a9e02d229R1421).

2) For `BUFFERED_RECORDS_PER_PARTITION_CONFIG`, similarly we use `@suppressWarning` on the code where we need to still reference it for backward compatibility like [here] (https://github.com/apache/kafka/pull/11424/files#diff-0e5e608831150c058e2ad1b45d38ad941739562588ec0fdb97cc9f742919fb1fR119) in TopologyConfig

Does that sound good to you? If yes we can go ahead and make it done.

Commit:
Adding suppresswarnings for usage of deprecated configs for backward compatibility

vamossagar12 commented:
> > hey @guozhangwang , did you get a chance to look at my above comment?
> 
> Ah yes, for StreamsConfigs in order for backward compatibility we still need to reference some deprecated configs. So:
> 
> 1. For all references of `CACHE_MAX_BYTES_BUFFERING_CONFIG`, we should be replacing it with the new config. We just add the `@suppressWarning` in the place we reference it only for [compatibility](https://github.com/apache/kafka/pull/11424/files#diff-07a8a87d3ee84c3e96918aa131033780cf4bc4d4a12ae376e91c059a9e02d229R1421).
> 2. For `BUFFERED_RECORDS_PER_PARTITION_CONFIG`, similarly we use `@suppressWarning` on the code where we need to still reference it for backward compatibility like [here] (https://github.com/apache/kafka/pull/11424/files#diff-0e5e608831150c058e2ad1b45d38ad941739562588ec0fdb97cc9f742919fb1fR119) in TopologyConfig
> 
> Does that sound good to you? If yes we can go ahead and make it done.

Thanks @guozhangwang . I added that annotation wherever applicable. Now the only thing pending is the renaming of the total-bytes metric :)

guozhangwang commented:
@vamossagar12 seems the jenkins fails still but due to `2 SpotBugs violations were found` this time.

ableegoldman commented:
@cadonna suggested renaming the metric to `input-buffer-bytes-total`, which makes sense to me as well

Commit:
Addressing spotBugsMain errors and renaming total-bytes to input-buffer-bytes-total

vamossagar12 commented:
@guozhangwang , i fixed that. Hopefully this time we won't see errors related to this PR. I will also check it. Also, @ableegoldman , thanks for confirming the metric name. I have updated the PR and also sent out an update on the KIP.

guozhangwang commented:
I'm fine with `input-buffer-bytes-total` too. Thanks!

BTW I thought we have a similar metric for store cache, but after checking I realize we actually do not have one yet :P Maybe we can add later.

guozhangwang commented:
There are still some failures in Jenkins, I'm gonna retrigger them.

If they still fails @vamossagar12 please take a look and see if they are relevant.

vamossagar12 commented:
> I'm fine with `input-buffer-bytes-total` too. Thanks!
> 
> BTW I thought we have a similar metric for store cache, but after checking I realize we actually do not have one yet :P Maybe we can add later.

Haha sure. Should I create another PR/JIRA for that?

> There are still some failures in Jenkins, I'm gonna retrigger them.
> 
> If they still fails @vamossagar12 please take a look and see if they are relevant.

vamossagar12 commented:
> There are still some failures in Jenkins, I'm gonna retrigger them.
> 
> If they still fails @vamossagar12 please take a look and see if they are relevant.

Thanks @guozhangwang . These are the tests that failed =>

```
Build / JDK 17 and Scala 2.13 / kafka.admin.LeaderElectionCommandTest.[1] Type=Raft, Name=testElectionResultOutput, Security=PLAINTEXT
Build / ARM / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldNotSendOldValuesOnMaterialization
Build / ARM / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldPassThroughOnMaterialization
Build / ARM / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldNotSendOldValuesOnMaterialization
Build / ARM / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldPassThroughOnMaterialization
Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.KStreamAggregationIntegrationTest.shouldCountUnlimitedWindows
Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.KStreamAggregationIntegrationTest.shouldCountUnlimitedWindows
Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldNotSendOldValuesOnMaterialization
Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldPassThroughOnMaterialization
Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldNotSendOldValuesOnMaterialization
Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldPassThroughOnMaterialization
Build / JDK 8 and Scala 2.12 / kafka.api.PlaintextAdminIntegrationTest.testReplicaCanFetchFromLogStartOffsetAfterDeleteRecords()
Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.KStreamAggregationIntegrationTest.shouldCountUnlimitedWindows
Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.KStreamAggregationIntegrationTest.shouldCountUnlimitedWindows
Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldNotSendOldValuesOnMaterialization
Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldPassThroughOnMaterialization
Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldNotSendOldValuesOnMaterialization
Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldPassThroughOnMaterialization
Build / JDK 17 and Scala 2.13 / kafka.api.PlaintextAdminIntegrationTest.testReplicaCanFetchFromLogStartOffsetAfterDeleteRecords()
Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.KStreamAggregationIntegrationTest.shouldCountUnlimitedWindows
Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.KStreamAggregationIntegrationTest.shouldCountUnlimitedWindows
Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldNotSendOldValuesOnMaterialization
Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldPassThroughOnMaterialization
Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldNotSendOldValuesOnMaterialization
Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.kstream.internals.KTableFilterTest.shouldPassThroughOnMaterialization
```

and these were the ones that had failed on my local as well. I am assuming these aren't relevant to this change. WDYT?

guozhangwang commented:
> and these were the ones that had failed on my local as well. I am assuming these aren't relevant to this change. WDYT?

Hmm.. do you mean these tests even fail on trunk? If they fail on your branch, and also here, they are likely to be relevant right?

Commit:
Fixing failing test cases plus corrected logic in ToplogyConfig for cache size config overrides

vamossagar12 commented:
> > and these were the ones that had failed on my local as well. I am assuming these aren't relevant to this change. WDYT?
> 
> Hmm.. do you mean these tests even fail on trunk? If they fail on your branch, and also here, they are likely to be relevant right?

@guozhangwang , sorry it was an oversight on my part. Looking at the names, it wasn't evident that these were related to this PR but apparently they were :D 

I ran the tests twice locally and there are 3-4 which fail intermittently due to timeout errors. When I ran them afterwards, even they seemed to have passed. Let's see what happens in this run.

guozhangwang commented:
> I ran the tests twice locally and there are 3-4 which fail intermittently due to timeout errors. When I ran them afterwards, even they seemed to have passed. Let's see what happens in this run.

Cool. Thanks @vamossagar12

vamossagar12 commented:
> > I ran the tests twice locally and there are 3-4 which fail intermittently due to timeout errors. When I ran them afterwards, even they seemed to have passed. Let's see what happens in this run.
> 
> Cool. Thanks @vamossagar12

Looks like the tests passed this time. ARM BUILD has some issue with 
`[2022-01-27T18:31:03.652Z] > Task :storage:unitTest FAILED` while other jdk based builds show the tests as green. Couple of examples of failed tests are =>

```
[2022-01-27T18:41:55.727Z] ConsumerBounceTest > testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup() FAILED
[2022-01-27T18:41:55.727Z]     org.opentest4j.AssertionFailedError: The remaining consumers in the group could not fetch the expected records
[2022-01-27T18:41:55.727Z]         at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:39)
[2022-01-27T18:41:55.727Z]         at org.junit.jupiter.api.Assertions.fail(Assertions.java:117)
[2022-01-27T18:41:55.727Z]         at kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup(ConsumerBounceTest.scala:331)
```
in jdk8 and 

```
[2022-01-27T19:49:17.737Z] > Task :core:integrationTest
[2022-01-27T19:49:17.737Z] kafka.api.PlaintextAdminIntegrationTest.testReplicaCanFetchFromLogStartOffsetAfterDeleteRecords() failed, log available in /home/jenkins/workspace/Kafka_kafka-pr_PR-11424/core/build/reports/testOutput/kafka.api.PlaintextAdminIntegrationTest.testReplicaCanFetchFromLogStartOffsetAfterDeleteRecords().test.stdout
[2022-01-27T19:49:17.737Z] 
[2022-01-27T19:49:17.737Z] PlaintextAdminIntegrationTest > testReplicaCanFetchFromLogStartOffsetAfterDeleteRecords() FAILED
[2022-01-27T19:49:17.737Z]     org.opentest4j.AssertionFailedError: Expected follower to catch up to log end offset 200
[2022-01-27T19:49:17.737Z]         at app//org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:39)
[2022-01-27T19:49:17.737Z]         at app//org.junit.jupiter.api.Assertions.fail(Assertions.java:117)
[2022-01-27T19:49:17.737Z]         at app//kafka.api.PlaintextAdminIntegrationTest.waitForFollowerLog$1(PlaintextAdminIntegrationTest.scala:730)
[2022-01-27T19:49:17.737Z]         at app//kafka.api.PlaintextAdminIntegrationTest.testReplicaCanFetchFromLogStartOffsetAfterDeleteRecords(PlaintextAdminIntegrationTest.scala:760)
``` 
in jdk 17. Do you think it's better now?

vamossagar12 commented:
Also, I created a new JIRA https://issues.apache.org/jira/browse/KAFKA-13624 for adding a new metric for cache size. I am assuming it needs a KIP and hence added needs-kip tag to it.

guozhangwang commented:
I checked the jenkins failures and they are irrelevant indeed this time. Merging to trunk now.

Thank you so much @vamossagar12 for the great contribution! Please feel free to update the JIRA ticket and the KIP wiki as well.

vamossagar12 commented:
> I checked the jenkins failures and they are irrelevant indeed this time. Merging to trunk now.
> 
> Thank you so much @vamossagar12 for the great contribution! Please feel free to update the JIRA ticket and the KIP wiki as well.

Thanks @guozhangwang and @ableegoldman  for the support on this one! Also, this ticket => https://issues.apache.org/jira/browse/KAFKA-13624, do you think it makes sense? Do we need a KIP for this?

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/namedtopology/TopologyConfig.java:
```diff
@@ -116,14 +123,30 @@ public TopologyConfig(final String topologyName, final StreamsConfig globalAppCo
+            }
         } else {
-            cacheSize = globalAppConfigs.getLong(CACHE_MAX_BYTES_BUFFERING_CONFIG);
+            cacheSize = globalAppConfigs.getLong(STATESTORE_CACHE_MAX_BYTES_CONFIG);
```
Hey guys, sorry I didn't get around to doing another full review of this earlier. This is a bug, we should always check whether the old deprecated config has been set and use that value if it was set and the new config was not. 

I noticed something similar in the TopologyTestDriver, although I'm not sure how much of an effect that would have since AFAICT there's not really "caching" in the TTD -- at least, there's not supposed to be according to the [javadocs](https://github.com/apache/kafka/pull/11424/files#diff-cd172cd5d1bf65c00d4c6f4c124f22c85473f6eed7715d4ac8c5a33092eb4695R158)

vamossagar12 replied:
@ableegoldman , I have added that check. The line above is in the else block when neither of the 2 configs are set. Here's the complete block of code =>
```
if (isTopologyOverride(STATESTORE_CACHE_MAX_BYTES_CONFIG, topologyOverrides) ||
                isTopologyOverride(CACHE_MAX_BYTES_BUFFERING_CONFIG, topologyOverrides)) {

            if (isTopologyOverride(STATESTORE_CACHE_MAX_BYTES_CONFIG, topologyOverrides) && isTopologyOverride(CACHE_MAX_BYTES_BUFFERING_CONFIG, topologyOverrides)) {
                cacheSize = getLong(STATESTORE_CACHE_MAX_BYTES_CONFIG);
                log.info("Topology {} is using both {} and deprecated config {}. overriding {} to {}",
                        topologyName,
                        STATESTORE_CACHE_MAX_BYTES_CONFIG,
                        CACHE_MAX_BYTES_BUFFERING_CONFIG,
                        STATESTORE_CACHE_MAX_BYTES_CONFIG,
                        cacheSize);
            } else if (isTopologyOverride(CACHE_MAX_BYTES_BUFFERING_CONFIG, topologyOverrides)) {
                cacheSize = getLong(CACHE_MAX_BYTES_BUFFERING_CONFIG);
                log.info("Topology {} is using deprecated config {}. overriding {} to {}", topologyName, CACHE_MAX_BYTES_BUFFERING_CONFIG, CACHE_MAX_BYTES_BUFFERING_CONFIG, cacheSize);
            } else {
                cacheSize = getLong(STATESTORE_CACHE_MAX_BYTES_CONFIG);
                log.info("Topology {} is overriding {} to {}", topologyName, STATESTORE_CACHE_MAX_BYTES_CONFIG, cacheSize);
            }
        } else {
            cacheSize = globalAppConfigs.getLong(STATESTORE_CACHE_MAX_BYTES_CONFIG);
        }
```

Am I missing something here?

ableegoldman replied:
Ah, I see the confusion. The `#isTopologyOverride` method checks whether the config has been overridden for the specific topology, ie has been set in the Properties passed in to `StreamsBuilder#build` -- it's not looking at what we call the `globalAppConfigs` which are the actual application configs: ie those passed in to the `KafkaStreams` constructor.

So basically there are two sets of configs. The value should be taken as the first of these to be set by the user, in the following order:
1) `statestore.cache.max.bytes` in `topologyOverrides` 
2) `cache.max.bytes.buffering` in `topologyOverrides` 
3)`statestore.cache.max.bytes` in `globalAppConfigs`
4) `cache.max.bytes.buffering` in `globalAppConfigs`

Essentially, using `#getTotalCacheSize` on the `topologyOverrides` if either of them is set (which this PR is doing) and on the `globalAppConfigs` if they are not (which is the regression here).

On that note -- we also need to move `##getTotalCacheSize` out of StreamsConfig, because it's a public class and wasn't listed as a public API in the KIP (nor should it be, imo). I recommend creating a new static utility class for things like this, eg `StreamsConfigUtils` in the `org.apache.kafka.streams.internals` package. There are some other methods that would belong there, for example the `StreamThread` methods `#processingMode` and `#eosEnabled` should be moved as well

Hope that all makes sense -- and lmk if you don't think you'll have the time to put out a full patch, and I or another Streams dev can help out 🙂

vamossagar12 replied:
thanks @ableegoldman! i am running slightly occupied currently. But, I will make the changes in the next few days. As you said, i will introduce a new utility class and move these methods out.

ableegoldman replied:
Sounds good! There's no rush, but I'll make sure we have your new PRs reviewed and merged quickly whenever they are ready, since you've worked so hard on this already. I'm sorry I wasn't able to make another pass on your original PR, but hopefully this won't be too much of a bother.

vamossagar12 replied:
Sorry just catching up on this comment @ableegoldman . That's perfectly fine!

ableegoldman commented on streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java:
```diff
@@ -329,7 +330,7 @@ private TopologyTestDriver(final InternalTopologyBuilder builder,
         final ThreadCache cache = new ThreadCache(
             logContext,
-            Math.max(0, streamsConfig.getLong(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG)),
+            Math.max(0, streamsConfig.getLong(StreamsConfig.STATESTORE_CACHE_MAX_BYTES_CONFIG)),
```
Ditto to [this comment](https://github.com/apache/kafka/pull/11424/files#r796139442), we can't just read off the new `statestore.cache.max.bytes` without checking whether the old `cache.max.bytes.buffering` was set

vamossagar12 replied:
Oh here, I thought since it's a test case so it shouldn't really matter. Isn't that the case?

ableegoldman replied:
> since it's a test case so it shouldn't really matter. Isn't that the case?

Well if someone is using the TTD to write unit tests and those tests start to fail after they upgrade because the caching is different, I would say that's compatibility change.

Although I read the TTD's javadocs earlier and remembered that it actually processes records synchronously, which effectively means that the only thing that matters/affects the TTD results is whether the cache size is non-zero or has been set to 0 -- and setting it to 0 only matters if it's correctly set to 0 in the TopologyConfig, not the value here. Which is a long way of saying that in hindsight, this config/bug doesn't really impact anything after all. In fact, imho we should probably just hard-code the TTD's ThreadCache cache size to 0 -- but let's not wrap that change into an already rather large PR in case there's something I'm not taking into account here.

So tl;dr, for future reference do still need to maintain backwards compatibility in the TTD since it's part of the public interface. But it just so happens that this particular bug doesn't actually break anything "real" or have any visible impact  (at least AFAICT

vamossagar12 replied:
got it.. If i understood correctly, we need a new PR with all the changes in this PR and the new ones along with document changes, right?

guozhangwang replied:
> got it.. If i understood correctly, we need a new PR with all the changes in this PR and the new ones along with document changes, right?

That sounds right to me.

mjsax commented:
Btw: this PR does not update the docs -- we should docs updates, too. (Also ok to do in a follow up PR.)

To ensure backward compatibility, it might also be good to split the actually PR into two: first so the internal config change without rewriting any tests -- this should ensure that no existing test breaks. -- In a follow up PR we update the test to use the new config.

And we should add a test that expliclity tests the old config to test backward compatibility explicitly.

mjsax commented:
I did not dig into the details myself. Anyway, might be better to discuss on the new PR?

ableegoldman commented:
> My read on the code is that we only need to change the TopologyTestDriver, while the first place seems fine to me. Did I miss anything?

If by "first place" you mean [the bug in the TopologyConfig class](https://github.com/apache/kafka/pull/11424#discussion_r796139442), then no, other way around actually. The TopologyConfig bug is the one that actually breaks compatibility, the TTD one actually doesn't really do anything -- if I understand the TTD correctly. See [this](https://github.com/apache/kafka/pull/11424#discussion_r798443526)

guozhangwang commented:
> If by "first place" you mean https://github.com/apache/kafka/pull/11424#discussion_r796139442, then no, other way around actually. The TopologyConfig bug is the one that actually breaks compatibility, the TTD one actually doesn't really do anything -- if I understand the TTD correctly. See https://github.com/apache/kafka/pull/11424#discussion_r798443526

Crystal! Thanks for the clarification @ableegoldman

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1034,14 +1037,15 @@ private static Metrics getMetrics(final StreamsConfig config, final Time time, f
+                // and then resize them later
+                streamThread = createAndAddStreamThread(0L, 0L, threadIdx);
+                final int numLiveThreads = getNumLiveStreamThreads();
+                resizeThreadCacheAndBufferMemory(numLiveThreads + 1);
```
One more thing -- @wcarlson5 mentioned there was an off-by-one error here due to the re-ordering of these calls. Specifically, the `+ 1` in this line was necessary before now because we called `resizeThreadCache` before actually adding the new thread, so we had to account for the new thread by adding one. But since we now create/add the new thread first, the `getNumLiveStreamThreads` method will actually return the correct number of threads, so we don't need the `+ 1` anymore.

On that note, I take it we reordered these calls because we now create the thread without the cache value and then call `resize` to set the cache after the thread has already been created. I was wondering: why do we need to do this post-construction resizing? I only looked at this part of the PR briefly, but it seems to me like we always have the actual cache size known when we're creating the thread, so can't we just pass that in to the StreamThread#create method/constructor? It's just a bit confusing to initialize the cache size to some random value, it took me a little while to figure out what was going on with that

guozhangwang replied:
> I was wondering: why do we need to do this post-construction resizing?

The main motivation is to consolidate the resizing of the thread cache and buffer within a single call. More details can be found in this comment thread: https://github.com/apache/kafka/pull/11424#discussion_r774058848 I suggested we initialize the new thread with 0 value -- should not be a random value? -- and then resize (at that time we have the correct number of threads to divide).

guozhangwang commented:
Thanks for keep looking into this issue, and continue to try getting it merged. After chatting with you offline I thought a bit more around the logic of pause / resume. And I think that besides the additional logic of:

1) pause all non-empty partitions when the global bytes threshold is exceeded;
2) resume all paused partitions when below the global bytes threshold.

We also do the following:

3) modify the single partition resume logic (a.k.a. https://github.com/apache/kafka/pull/11424/files#diff-a76674468cda8772230fb8411717cf9068b1a363a792f32c602fb2ec5ba9efd7R722) that if the partition buffer becomes empty after the last record is retrieved, resume it if it was paused. I.e. we should remove the TODO since it would not be removed, and also not condition on `maxBufferedSize != -1`.
4) deprecated the single partition pause logic (a.k.a. https://github.com/apache/kafka/pull/11424/files#diff-a76674468cda8772230fb8411717cf9068b1a363a792f32c602fb2ec5ba9efd7R977) just like what you did previously still, since we would still remove it in the future.

cc @mjsax @vvcephei who also looked into this.
